File: Java/ContainerWithMostWater.java
//brute force O(n^2) method 
class Solution {
    public int maxArea(int[] height) {
        // brute force
        int max = 0;

        for (int i = 0; i < height.length - 1; i++) {
            for (int j = 1; j < height.length; j++) {
               int width = j - i;
               int length = Math.min(height[i], height[j]);
               int area = width * length;
               if (area > max) max = area;
            }            
        }
        return max;

    }
}

//Optimal solution  O (n) two pointer solution
class Solution {
    public int maxArea(int[] height) {
        int max = 0;

        int l = 0, r = height.length - 1;
        while (l < r) {
            int area = Math.min(height[l], height[r]) * (r - l) ; // bottleneck times the difference between l and r
            max = Math.max(area, max);
            if (height[l] > height[r])
                r--;
            else
                l++;
        }

        return max;
    }


}

File: Python/Two Pointers/containerWithMostWater.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:

        l, r = 0, len(height) - 1
        maxL, maxR = height[l], height[r]
        maxA = min(height[l], height[r]) * (r - l)

        while l < r:
            # limiting factor will be height so i'd say we increase or decrease based on which pointer is shorter
            # only check when one of the heights increases 
            if height[l] <= height[r]:
                #increase l
                l += 1
            elif height[l] > height[r]:
                # increase r
                r -= 1

            # if the height has increased, we can do a check
            #skip any time the height does not increase because we are going inwards
            # this will make area less if height does not increase
            if maxL < height[l]:
                maxL = height[l]
                # get max between current area and max area
                maxA = max(maxA, min(height[l], height[r]) * (r - l))
            elif maxR < height[r]:
                maxL = height[l]
                # get max between current area and max area
                maxA = max(maxA, min(height[l], height[r]) * (r - l))
      

        return maxA